File: web_iot/python/led.py
# pip install pyserial flask

import serial
import flask
import json

app = flask.Flask(__name__, static_folder="../static")
dev_name = "/dev/tty.usbmodem14201"
speed = 9600
port = serial.Serial(dev_name, speed)
print("Serial device ready :", port.isOpen())

@app.route("/")
def index():
    return flask.redirect("/static/led.html")

@app.route("/led_on")
def ledOn():
    port.write("1".encode())
    ret = port.readline()
    return ret

@app.route("/led_off")
def ledOff():
    port.write("0".encode())
    ret = port.readline()
    return ret

@app.route('/operate/<cmd>')
def led(cmd):
    val = flask.request.values['val']   
    if cmd == "led":
        val = flask.request.values['val']   
        print('/operate/', cmd, val) 
        if val == 'on':
            port.write("1".encode())
        elif val == 'off':
            port.write("0".encode())
        msg = port.readline().decode()
        ret = {"code": False, "message":msg }
    else :
        ret = {"code": False, "message":"Unknown Command" }
    print(ret)
    return json.dumps(ret)

if __name__ == '__main__':
    app.run(host="0.0.0.0", port=8080)

File: web_iot/nodeJs/led.js
//npm install express serialport
const express = require('express');
const SerialPort = require("serialport");
const readline = require('readline');

let app = express();

var dev_name = "/dev/tty.usbmodem14201"
var speed = 9600;
let port_r;
var port = new SerialPort(dev_name, {baudRate:speed}, (err)=>{
    if(err) console.log("serial port open err", err);
    else port_r = readline.createInterface(port);
});


app.get('/', (req, res)=>{
    res.redirect('/led.html');
});
app.use(express.static(__dirname + '/../static/'));

app.get('/led_on', (req, res)=>{
    if(port.isOpen){
        port_r.once('line', (data)=>{
            console.log(data);
            res.send(data);
        });
        port.write('1');
    }
});
app.get('/led_off', (req, res)=>{
    if(port.isOpen){
        port_r.once('line', (data)=>{
            console.log(data);
            res.send(data);
        });
        port.write('0')
    }
});

app.get('/operate/:cmd', (req, res)=>{
    if(req.params.cmd === 'led'){
        console.log(req.url, req.params.cmd, req.query.val)
        if(port.isOpen){
            port_r.once('line', (data)=>{
                console.log(data);
                res.json({"code": false, "message":data })
            });
            if(req.query.val === 'on') port.write('1');
            else if(req.query.val === 'off') port.write('0');
            else res.json({code : false,  message : 'Unknown Command' })
        }
    
    }
});

app.listen(8080, ()=>{
    console.log("server on 8080.");
});
